Amplify/Categories/DataStore/Model/Temporal/Date.swift

```diff
@@ -0,0 +1,49 @@
+//
+// Copyright 2018-2020 Amazon.com,
+// Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+extension Temporal {
+
+    /// An extension that makes the `Date` struct conform with the `TemporalSpec` protocol.
+    /// When used in persistence operations, the granularity of the different date representations
+    /// is set by using different scalar types: `Date`, `DateTime` and `Time`.
+    ///
+    /// In those scenarios, the standard `Date` is formatted to ISO-8601 without the time.
+    /// When the full date information is required, use `DateTime` instead.
+    ///
+    public struct Date: TemporalSpec, DateUnitOperable {
+
+        public static func now() -> Self {
+            return Temporal.Date(Foundation.Date())
+        }
+
+        public let date: Foundation.Date
+
+        public init(iso8601String: String) throws {
+            guard let date = Temporal.Date.iso8601Date(from: iso8601String) else {
+                throw DataStoreError.invalidDateFormat(iso8601String)
+            }
+            self.date = date
+        }
+
+        public init(_ date: Foundation.Date) {
+            // sets the date to a fixed instant so date-only operations are safe
+            let calendar = Date.iso8601Calendar
+            var components = calendar.dateComponents(
+                [.year, .month, .day, .hour, .minute, .second, .nanosecond, .timeZone],
+                from: date
+            )
+            // reset time to a fixed time
+            components.hour = 12
```

Comment:
This was a tough decision. The reason I didn't use `startOfDay` from `Calendar` was due to previous issues with Daylight Saving Time during operations.

In your example I would consider that a programming mistake, since `Temporal.Date` is used to represent date-only values so 4am does not fit in that context, the operations `date + .hours(4)` where date is a `Temporal.Date` doesn't even compile. `Temporal.DateTime` would be the correct type to use.

That said, I think it's fair to assume it can happen if customers access the underlying `Foundation.Date` object and do their operations. Thoughts on that? My guess is that we should properly document our implementation but not try to support time operation on Date-only types. For instance, if you parse `HH:mm` to a `Foundation.Date`, Swift sets the date to `2000-01-01`, so I guess those compromises are fine.

Reply:
Yes, I think there would be confusion if customers operate on the `foundationDate`. If we don't want them to do that, we should make it private. "I store the blog's future 'publish date' in my GraphQL backend, which is used to run a cron job, so I'll just pass it to my underlying system like: `startPublishJob(at: blog.publishDate.foundationDate)`"

Additionally:

```swift
let calendar = Calendar(identifier: .gregorian)
let components = DateComponents(year: 2020, month: 5, day: 20)
calendar.date(from: components)! // "May 20, 2020 at 12:00 AM"
```

I think we're violating principle of least surprise by picking other than the start-of-day